### PR TITLE
oshmem: Fix unsigned type comparisons

### PR DESCRIPTION
**Pull request by gleon99**
Fix unsigned comparisons by using the correct type casting

### CODE REVIEW

**Comment by awlauria** [comment]
ok to test

File: oshmem/shmem/c/shmem_wait.c [hunk 1]
```diff
@@ -117,13 +117,13 @@ SHMEM_TYPE_WAIT_UNTIL(_int, volatile int, SHMEM_INT, shmem)
 SHMEM_TYPE_WAIT_UNTIL(_long, volatile long, SHMEM_LONG, shmem)
 SHMEM_TYPE_WAIT_UNTIL(_longlong, volatile long long, SHMEM_LLONG, shmem)
 SHMEM_TYPE_WAIT_UNTIL(_ushort, volatile unsigned short, SHMEM_SHORT, shmem)
-SHMEM_TYPE_WAIT_UNTIL(_uint, volatile unsigned int, SHMEM_INT, shmem)
-SHMEM_TYPE_WAIT_UNTIL(_ulong, volatile unsigned long, SHMEM_LONG, shmem)
-SHMEM_TYPE_WAIT_UNTIL(_ulonglong, volatile unsigned long long, SHMEM_LLONG, shmem)
+SHMEM_TYPE_WAIT_UNTIL(_uint, volatile unsigned int, SHMEM_UINT, shmem)
+SHMEM_TYPE_WAIT_UNTIL(_ulong, volatile unsigned long, SHMEM_ULONG, shmem)
+SHMEM_TYPE_WAIT_UNTIL(_ulonglong, volatile unsigned long long, SHMEM_ULLONG, shmem)
```
Comment (hoopoepg) [inline oshmem/shmem/c/shmem_wait.c:122]: SHMEM_USHORT is missing?

**Review by devreal — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

File: oshmem/shmem/c/shmem_wait.c (hunk 1 above)
Comment (yosefe) [reply oshmem/shmem/c/shmem_wait.c:122]: @gleon99 line 119 should use SHMEM_USHORT instead of SHMEM_SHORT

**Review by yosefe — CHANGES_REQUESTED** [review]
Should we also fix usage of SHMEM_TYPE_TEST macro, and shmem_wait / SHMEM_TYPE_WAIT ?

File: oshmem/mca/spml/base/spml_base.c [hunk 2]
```diff
@@ -80,41 +96,97 @@ int mca_spml_base_test(void* addr, int cmp, void* value, int datatype, int *out_
         SPML_BASE_DO_CMP((*out_value), int_addr, cmp, int_value);
         break;
 
+        /* Unsigned Int */
+    case SHMEM_UINT:
+        uint_value = *(unsigned int*) value;
+        uint_addr = (unsigned int*) addr;
+        SPML_BASE_DO_CMP((*out_value), uint_addr, cmp, uint_value);
+        break;
```
Comment (yosefe) [inline oshmem/mca/spml/base/spml_base.c:104]: maybe we can define macro like this
```
#define SPML_BASE_TEST_CASE(_type, _shmem_type, _addr, _value, _cmp, _out_value) \
    case _shmem_type: \
        { \
            _type typed_value = *(const _type*)value; \
            const _type *typed_addr = (const _type*)addr; \
            SPML_BASE_DO_CMP((*_out_value), typed_addr , _cmp, typed_value); \
        } \
        break; \
```

and similarly for wait, wait_until
there is too much code duplication and hard to notice issues..

File: oshmem/shmem/c/shmem_wait.c [hunk 3]
```diff
@@ -89,11 +89,18 @@
...
 SHMEM_TYPE_WAIT(_long, volatile long, SHMEM_LONG, shmem)
+SHMEM_TYPE_WAIT(_ulong, volatile unsigned long, SHMEM_ULONG, shmem)
 SHMEM_TYPE_WAIT(_longlong, volatile long long, SHMEM_LLONG, shmem)
+SHMEM_TYPE_WAIT(_ulonglong, volatile unsigned long long, SHMEM_LLONG, shmem)
 SHMEM_TYPE_WAIT(_int32, int32_t, SHMEM_INT32_T, shmemx)
-SHMEM_TYPE_WAIT(_int64, int64_t, SHMEM_INT64_T, shmemx)
+SHMEM_TYPE_WAIT(_uint32, uint32_t, SHMEM_UINT32_T, shmemx)
+SHMEM_TYPE_WAIT(_uint64, uint64_t, SHMEM_UINT64_T, shmemx)
+SHMEM_TYPE_WAIT(_size, size_t, SHMEM_SIZE_T, shmem)
+SHMEM_TYPE_WAIT(_ptrdiff, ptrdiff_t, SHMEM_PTRDIFF_T, shmem)
```
Comment (yosefe) [inline oshmem/shmem/c/shmem_wait.c:103]: actually for unsigned types, shmem_wait is not defined, so not needed
https://github.com/open-mpi/ompi/blob/ef7ab9bf9699fa81ba27aed7a32305e9f3e313c4/oshmem/include/shmem.h.in#L1639-L1643

Comment (gleon99) [reply oshmem/shmem/c/shmem_wait.c:103]: Isn't it a problem by itself?

Comment (yosefe) [reply oshmem/shmem/c/shmem_wait.c]: it's not required by openshmem spec
http://www.openshmem.org/site/sites/default/site_files/OpenSHMEM-1.5.pdf

**Review by yosefe — APPROVED** [review]

**Pull request merged**